File: org.palladiosimulator.blockchainsystems.plugin/src/main/java/org/palladiosimulator/blockchainsystems/plugin/logging/LogOutputProviderImpl.java
package org.palladiosimulator.blockchainsystems.plugin.logging;

import java.util.HashSet;
import java.util.Set;

import org.palladiosimulator.blockchainsystems.core.simulation.logoutputs.abstractions.LogOutputProvider;
import org.palladiosimulator.blockchainsystems.core.tracing.TraceEventLogOutput;
import org.palladiosimulator.blockchainsystems.loggers.*;

/**
 * Implementation of the {@link LogOutputProvider} interface.
 * Supports console, file, and database logging.
 *
 * @author Yannik Sproll
 */
public class LogOutputProviderImpl implements LogOutputProvider {

    private final boolean _useConsoleLogging;

    private final boolean _useFileLogging;
    private final String _fileLoggingDirectoryPath;

    private final boolean _useDatabaseLogging;
    private final String _dbServer;
    private final String _dbName;
    private final int _dbPort;
    private final String _dbUsername;
    private final String _dbPassword;

    public LogOutputProviderImpl(
            boolean useConsoleLogging,
            boolean useFileLogging,
            String fileLoggingDirectoryPath,
            boolean useDatabaseLogging,
            String dbServer,
            int dbPort,
            String dbName,
            String dbUsername,
            String dbPassword
    ) {
        _useConsoleLogging = useConsoleLogging;
        _useFileLogging = useFileLogging;
        _fileLoggingDirectoryPath = fileLoggingDirectoryPath;
        _useDatabaseLogging = useDatabaseLogging;
        _dbServer = dbServer;
        _dbName = dbName;
        _dbPort = dbPort;
        _dbUsername = dbUsername;
        _dbPassword = dbPassword;
    }


    public Set<TraceEventLogOutput> getLogOutputs() {
        HashSet<TraceEventLogOutput> logOutputs = new HashSet<TraceEventLogOutput>();

        if (_useConsoleLogging) {
            logOutputs.add(createConsoleLogger());
        }

        if (_useFileLogging) {
            logOutputs.add(createFileLogger());
        }

        if (_useDatabaseLogging) {
            logOutputs.add(createDatabaseLogger());
        }

        return logOutputs;
    }

    private TraceEventLogOutput createConsoleLogger() {
        return new TraceEventConsoleLogger();
    }

    private TraceEventLogOutput createFileLogger() {
        return new TraceEventFileLogger(_fileLoggingDirectoryPath);
    }

    private TraceEventLogOutput createDatabaseLogger() {
        return new TraceEventPostgresDbLogger(
                _dbServer,
                _dbPort,
                _dbName,
                _dbUsername,
                _dbPassword
        );
    }
}

File: org.palladiosimulator.blockchainsystems.plugin/src/main/kotlin/org/palladiosimulator/blockchainsystems/plugin/logging/LogOutputProviderImpl.kt
package org.palladiosimulator.blockchainsystems.plugin.logging

import kotlinx.serialization.json.Json
import org.palladiosimulator.blockchainsystems.core.simulation.logoutputs.abstractions.LogOutputProvider
import org.palladiosimulator.blockchainsystems.core.tracing.TraceEventLogOutput
import org.palladiosimulator.blockchainsystems.loggers.TraceEventConsoleLogger
import org.palladiosimulator.blockchainsystems.loggers.TraceEventFileLogger
import org.palladiosimulator.blockchainsystems.loggers.TraceEventPostgresDbLogger

/**
 * Implementation of the [LogOutputProvider] interface.
 * Supports console, file, and database logging.
 *
 * @author Yannik Sproll, Davis Riedel
 */
class LogOutputProviderImpl(
  private val jsonSerializer: Json,
  private val useConsoleLogging: Boolean,
  private val useFileLogging: Boolean,
  private val fileLoggingDirectoryPath: String,
  private val useDatabaseLogging: Boolean,
  private val dbServer: String,
  private val dbPort: String,
  private val dbName: String,
  private val dbUsername: String,
  private val dbPassword: String
) : LogOutputProvider {

  override val logOutputs: MutableSet<TraceEventLogOutput>
    get() {
      val logOutputs = HashSet<TraceEventLogOutput>()

      if (useConsoleLogging) {
        logOutputs.add(createConsoleLogger())
      }

      if (useFileLogging) {
        logOutputs.add(createFileLogger())
      }

      if (useDatabaseLogging) {
        logOutputs.add(createDatabaseLogger())
      }

      return logOutputs
    }

  private fun createConsoleLogger(): TraceEventLogOutput {
    return TraceEventConsoleLogger(jsonSerializer)
  }

  private fun createFileLogger(): TraceEventLogOutput {
    return TraceEventFileLogger(
      jsonSerializer,
      fileLoggingDirectoryPath
    )
  }

  private fun createDatabaseLogger(): TraceEventLogOutput {
    return TraceEventPostgresDbLogger(
      dbServer,
      dbPort,
      dbName,
      dbUsername,
      dbPassword
    )
  }
}